[test] Add rom_e2e_flash_ctrl_init test.

jadephilipoom opened the pull request:
Add the `rom_e2e_flash_ctrl_init` test (as described in the ROM e2e testplan). Resolves https://github.com/lowRISC/opentitan/issues/14500

Note: in the `rom_e2e_testplan.hjson`, the `keymgr_init` e2e test was actually already under the `flash_ctrl` testplan entry; I assumed this was a mistake and moved it to the `keymgr_init` entry just below.

jadephilipoom commented on sw/device/silicon_creator/rom/e2e/rom_e2e_flash_ctrl_init_test.c:
```diff
@@ -0,0 +1,304 @@
+  // * ERASE_EN  = false
+  uint32_t info_page_cfg =
+      abs_mmio_read32(kBase + FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_REG_OFFSET);
+  // TODO: the below fails because enable is false, but it's unclear from the
+  // flash documentation if this actually disables all the other non-enabled
+  // things. Need to investigate.
+  // CHECK(kMultiBitBool4True ==
+  //      bitfield_field32_read(info_page_cfg,
+  //                            FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_EN_1_FIELD));
```
This is something I wasn't sure about when reading the docs and writing the test. The documentation for `flash_ctrl` says that if the `en` field is set then "the following fields apply": https://docs.opentitan.org/hw/ip/flash_ctrl/doc/#Reg_bank0_info0_page_cfg_0

To me this wording kind of implies that if `en` is unset, the fact that read/prog/erase is disabled in the following fields has no effect and ...some other configuration applies. Is that correct? If so, the ROM driver should probably be modified to set the `en` flag.

@tjaychen for `flash_ctrl` questions, since I see you have a lot of commit history in `hw/ip/flash_ctrl/rtl` :slightly_smiling_face:

tjaychen replied:
yeah that's right. At least what you describe is how it should behave. Do let me know if that's not what you see. I'll make the wording more explicit as well.

The `en` concept doesn't make quite as much sense to `info pages`, it just gives you a quick knob to disable all the settings should you need it. It applies a bit more for the `data pages`, because it gives you a way to say I want the controls to fall to the "default'.  The info pages don't really have a default concept right now (in hindsight, maybe I should have added that).

alphan replied:
No access to the info page if the corresponding `EN` is `false`, right @tjaychen ?

jadephilipoom replied:
Okay, so if everything is disabled, the `en` bit doesn't matter because the "default" in this case is that everything is disabled.  In this case I think we should modify the code anyway to set `en`, just for clarity/in case the info pages do get a default value later on. I can take care of this change if @alphan agrees.

alphan replied:
@jadephilipoom which code are you referring to? We set those registers in `flash_ctrl_init()`. Please note that these bitfields use mubi variants and they are disabled unless they are set to true.

jadephilipoom commented on sw/device/silicon_creator/rom/e2e/rom_e2e_flash_ctrl_init_test.c:
```diff
@@ -0,0 +1,304 @@
+ * Verify that the flash controller is initialized.
+ */
+static void is_initialized_test(void) {
+  // Check that the physical status register does not say flash controller is
+  // still initializing.
+  uint32_t phy_status =
+      abs_mmio_read32(kBase + FLASH_CTRL_PHY_STATUS_REG_OFFSET);
+  CHECK(bitfield_bit32_read(phy_status, FLASH_CTRL_PHY_STATUS_INIT_WIP_BIT) ==
+        false);
+
+  // Check that the general status register does not say flash controller is
+  // still initializing, and that the FIFOs are empty.
+  uint32_t status = abs_mmio_read32(kBase + FLASH_CTRL_STATUS_REG_OFFSET);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_INIT_WIP_BIT) == false);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_RD_EMPTY_BIT) == true);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_PROG_EMPTY_BIT) == true);
```
The `init_wip` bits were the best way I could determine from the docs to check if flash was initialized. However, these are false at reset, so if some failure caused the initialization to never start then this test wouldn't catch it. I'd be interested to know if there's a better way to check this.

tjaychen replied:
hmm... that's good feedback, I probably should have had an actual status that just indicated whether we've initialized. 
I can add that if we think it helps? 

Without that, the best I could suggest is that when you initialize, first poll for wip to become high, then check for it to be false later.

alphan replied:
Would the following work?
```
(pseudo-code)
Verify that flash_ctrl.INIT.VAL == 1 // rw1s, requested init.
Verify that INIT_WIP == 0 // ro, 0 if no need to panic
```

jadephilipoom replied:
This works, perfect!

tjaychen commented:
one last thing, regarding a lot of tests for `flash_init` and `keymgr`, you guys are probably going to be impacted by this PR whenever it lands. See #15776

Maybe we should discuss a bit on how we should coordinate this change.

alphan commented:
> Note: in the `rom_e2e_testplan.hjson`, the `keymgr_init` e2e test was actually already under the `flash_ctrl` testplan entry; I assumed this was a mistake and moved it to the `keymgr_init` entry just below.

Thanks @jadephilipoom! Yes, looks like it was a mistake.

alphan commented on sw/device/silicon_creator/rom/e2e/BUILD:
```diff
@@ -72,6 +72,29 @@ SLOTS = {
+opentitan_functest(
+    name = "rom_e2e_flash_ctrl_init",
+    srcs = ["rom_e2e_flash_ctrl_init_test.c"],
+    dv = dv_params(
+        rom = "//sw/device/silicon_creator/rom",
+    ),
+    signed = True,
+    targets = [
+        "dv",
+        "cw310_rom",
+        "verilator",
+    ],
+    verilator = verilator_params(
+        timeout = "eternal",
+        rom = "//sw/device/silicon_creator/rom",
+    ),
```
```suggestion
    signed = True,
    targets = [
        "cw310_rom",
    ],
```

alphan commented on sw/device/silicon_creator/rom/e2e/rom_e2e_flash_ctrl_init_test.c:
```diff
@@ -0,0 +1,298 @@
+  uint32_t data_addr = (0 * FLASH_CTRL_PARAM_BYTES_PER_BANK) +
+                       (1 * FLASH_CTRL_PARAM_BYTES_PER_BANK);
+
+  // Attempt to read, expect an error.
+  uint32_t read_data[1];
+  CHECK(flash_ctrl_data_read(data_addr, 1, read_data) ==
+        kErrorFlashCtrlDataRead);
+
+  // Attempt to write, expect an error.
+  uint32_t write_data[1] = {0};
+  CHECK(flash_ctrl_data_write(data_addr, 1, write_data) ==
+        kErrorFlashCtrlDataWrite);
+
+  // Attempt to erase, expect an error.
+  CHECK(flash_ctrl_data_erase(data_addr, kFlashCtrlEraseTypePage) ==
+        kErrorFlashCtrlDataErase);
```
Thanks for checking that all these operations fail too!

alphan reviewed (approved):
Thanks @jadephilipoom !

We should also iterate over different OTP configurations but we can do that after merging this PR and `.../rom/e2e/BUILD` has more examples for OTP splicing.

jadephilipoom commented:
> one last thing, regarding a lot of tests for `flash_init` and `keymgr`, you guys are probably going to be impacted by this PR whenever it lands. See #15776
> 
> Maybe we should discuss a bit on how we should coordinate this change.

What exactly is the scope of this change? If it's just changing the reset values then it shouldn't impact this test, at least -- the flash controller doesn't seem to rely on the reset values anywhere.

alphan commented:
> > one last thing, regarding a lot of tests for `flash_init` and `keymgr`, you guys are probably going to be impacted by this PR whenever it lands. See #15776
> > Maybe we should discuss a bit on how we should coordinate this change.
> 
> What exactly is the scope of this change? If it's just changing the reset values then it shouldn't impact this test, at least -- the flash controller doesn't seem to rely on the reset values anywhere.

Please see @tjaychen's clarification [here](https://github.com/lowRISC/opentitan/pull/15776#issuecomment-1293970788). I don't think this PR is affected since we are not working with concrete expected values for keymgr outputs.

jadephilipoom commented:
Summarizing an offline conversation with @alphan -- I now removed all checks related to `kFlashCtrlInfoPageCreatorSecret` because as of #15147 the lockdown of creator-secret pages is in ROM_EXT, not ROM. I also removed that check from the list in the testplan HJSON file.

alphan commented on sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson:
```diff
@@ -1089,12 +1089,11 @@
       tags: ["rom", "no_dv", "fpga", "silicon"]
       stage: V2
-      tests: ["rom_keymgr_init"]
+      tests: ["rom_e2e_flash_ctrl_init"]
```
@timothytrippel we should leave this empty, otherwise it will show up in the dv dashboard, right?

timothytrippel replied:
this will have no effect since any test with the `no_dv` tag will not end up in the dashboard; only if there are plans to add it to dv, then we should leave it blank until its added to DV.

timothytrippel replied:
though since it has no impact, it would be cleaner / more consistent if this was empty

alphan commented on sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson:
```diff
@@ -1112,7 +1111,7 @@
       tags: ["rom", "no_dv", "fpga", "silicon"]
       stage: V2
-      tests: []
+      tests: ["rom_keymgr_init"]
```
@timothytrippel  this one too?

timothytrippel replied:
same as above

alphan reviewed (approved):
Thanks @jadephilipoom
